lib/RouterMixin.js

```diff
@@ -83,11 +83,36 @@ RouterMixin = {
         }
 
         self._historyListener = function (e) {
+            debug('history listener invoked', e, url, self.state.route.url);
+
             if (context) {
+                var state = self.state || {};
                 var url = self._history.getUrl();
-                debug('history listener invoked', e, url, self.state.route.url);
-                if (url !== self.state.route.url) {
-                    context.executeAction(navigateAction, {type: TYPE_POPSTATE, url: url, params: (e.state && e.state.params)});
+                var currentUrl = state.route.url;
+                var route = state.route || {};
+                var onBeforeUnloadText = typeof window.onbeforeunload === 'function' ? window.onbeforeunload() : '';
+                var confirmResult = onBeforeUnloadText ? window.confirm(onBeforeUnloadText) : true;
+                var nav = route.navigate || {};
+                var navParams = nav.params || {};
+                var enableScroll = self._enableScroll && nav.preserveScrollPosition;
+                var historyState = {
+                    params: (nav.params || {}),
+                    scroll: {
+                        x: (enableScroll ? window.scrollX : 0),
+                        y: (enableScroll ? window.scrollY : 0)
+                    }
+                };
+                var pageTitle = navParams.pageTitle || null;
+
+                if (!confirmResult && currentUrl) {
```

Comment:
I don't think `currentUrl` would be undefined or empty.  This `&& currentUrl` adds confusion to the if/else logic, since it does not really make sense to continue with navigation if user has said no (i.e. !confirmResult).  I would just remove `&& currentUrl`.  

If you are worried about it being falsy, I'd rather check for it inside the if clause and throw an error, since it is not supposed to be that way.